src/context/internal.rs

```diff
@@ -116,6 +129,96 @@ pub trait ContextInternal<'a>: Sized {
 
     fn activate(&self) { self.scope_metadata().active.set(true); }
     fn deactivate(&self) { self.scope_metadata().active.set(false); }
+
+    #[cfg(feature = "legacy-runtime")]
+    fn try_catch_internal<'b: 'a, F>(&mut self, f: F) -> Result<Handle<'a, JsValue>, Handle<'a, JsValue>>
+        where F: UnwindSafe + FnOnce(&mut Self) -> JsResult<'b, JsValue>
+    {
+        // A closure does not have a guaranteed layout, so we need to box it in order to pass
+        // a pointer to it across the boundary into C++.
+        let rust_thunk = Box::into_raw(Box::new(f));
+
+        let mut local: MaybeUninit<raw::Local> = MaybeUninit::zeroed();
+        let mut unwind_value: MaybeUninit<*mut c_void> = MaybeUninit::zeroed();
+
+        let ctrl = unsafe {
+            neon_runtime::try_catch::with(try_catch_glue::<Self, F>,
+                                          rust_thunk as *mut c_void,
+                                          (self as *mut Self) as *mut c_void,
+                                          local.as_mut_ptr(),
+                                          unwind_value.as_mut_ptr())
+        };
+
+        match ctrl {
+            TryCatchControl::Panicked => {
+                let unwind_value: Box<dyn Any + Send> = *unsafe {
+                    Box::from_raw(unwind_value.assume_init() as *mut Box<dyn Any + Send>)
+                };
+                resume_unwind(unwind_value);
+            }
+            TryCatchControl::Returned => {
+                let local = unsafe { local.assume_init() };
+                Ok(JsValue::new_internal(local))
+            }
+            TryCatchControl::Threw => {
+                let local = unsafe { local.assume_init() };
+                Err(JsValue::new_internal(local))
+            }
+        }
+    }
+
+    #[cfg(feature = "napi-runtime")]
+    fn try_catch_internal<'b: 'a, F>(&mut self, f: F) -> Result<Handle<'a, JsValue>, Handle<'a, JsValue>>
+        where F: UnwindSafe + FnOnce(&mut Self) -> JsResult<'b, JsValue>
+    {
+        let result = f(self);
+        let mut local: MaybeUninit<raw::Local> = MaybeUninit::zeroed();
+        unsafe {
+            if neon_runtime::error::catch_error(self.env().to_raw(), local.as_mut_ptr()) {
+                Err(JsValue::new_internal(local.assume_init()))
+            } else if let Ok(result) = result {
+                Ok(result)
+            } else {
+                Ok(JsUndefined::new_internal(self.env()).upcast())
```

Comment:
Shouldn't this be `Err`?

Reply:
Check my logic: `napi::check_and_clear_last_exception()` only produces an exception if the VM was in the throwing state. So in this branch, we aren't in a throwing state, but the Neon user returned `Err(Throw)`—in other words, this is the case where the Neon user incorrectly used the `NeonResult` protocol.

My reasoning was that `cx.try_catch()` should correctly indicate that nothing was caught even if the Neon user was incorrectly using the protocol.

As I thought about how dirty I felt for silently swallowing an error and replacing it with `undefined`, it occurred to me just now that maybe a better route is to panic.

Reply:
In order to support a generic `T: Value` on this, this either needs to be an `Err` or a `panic!`. Panic makes more sense to me since it's not the intended use.